### PR TITLE
Raindrops

### PR DESCRIPTION
**Pull request by amymhaddad**
@paulghaddad Pls review the solution to this problem and the tests I wrote for it.

### CODE REVIEW

File: raindrops/raindrops.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+FACTORS_TO_SOUNDS = {
+    3: "Pling",
+    5: "Plang",
+    7: "Plong"
+}
+
+
+
+def factors(n):
```
Comment (paulghaddad) [inline raindrops/raindrops.py]: Nice and clean algorithm!

File: raindrops/raindrops.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        for inner_number in range(1, n+1):
+            if outer_number * inner_number == n:
+                factors.append(outer_number)
+    return factors
+
+
+def raindrops(n):
+    """Return a sound based on a number's factors"""
+
+    list_of_factors = factors(n)
```
Comment (paulghaddad) [inline raindrops/raindrops.py]: Good job extracting `factors` to its own function.

File: raindrops/raindrops.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    if len(list_of_factors) == 0:
+        return ""
+    
+    if not common_factors(n):
+        return str(n)        
+
+    return "".join([FACTORS_TO_SOUNDS[number] for number in list_of_factors if number in FACTORS_TO_SOUNDS])
```
Comment (paulghaddad) [inline raindrops/raindrops.py]: The P really knows her comprehensions! 👏

File: raindrops/raindrops.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+def common_factors(n):
+    factors_to_sounds_factors = set(FACTORS_TO_SOUNDS.keys())
```
Comment (paulghaddad) [inline raindrops/raindrops.py]: Love the use of `FACTORS_TO_SOUNDS`.

File: raindrops/raindrops.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    number_factors = set(factors(n))
```
Comment (paulghaddad) [inline raindrops/raindrops.py]: `factors_of_n` may be a better variable name.

File: raindrops/raindrops.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    return factors_to_sounds_factors.intersection(number_factors)
```
Comment (paulghaddad) [inline raindrops/raindrops.py]: Excellent!

File: raindrops/test_raindrops.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+from raindrops import raindrops
+
+
+
+def test_factor_3():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: Let's discuss these tests in person. They look good, but some thoughts around naming that are too long to write.

Comment (paulghaddad) [reply raindrops/test_raindrops.py]: Such as `test_number_with_factor_3`

`test_number_with_factor_3_and_5`

`test_number_with_no_matching_factors`

File: raindrops/test_raindrops.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    assert raindrops(3) == "Pling"
+
+def test_factor_5():
+    assert raindrops(5) == "Plang"
+
+def test_factor_7():
+    assert raindrops(7) == "Plong"
+
+def test_factors_for_number_28():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: `test_number_multiple_of_7`

File: raindrops/test_raindrops.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+def test_factors_3_and_5():
+    assert raindrops(15) == "PlingPlang"
+
+def test_factors_3_and_7():
+    assert raindrops(21) == "PlingPlong"
+
+def test_factors_5_and_7():
+    assert raindrops(35) == "PlangPlong"
+
+def test_factors():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: What does this test? Naming...

File: raindrops/test_raindrops.py (hunk 7 above)
Comment (paulghaddad) [reply raindrops/test_raindrops.py]: Also try to group the tests...which you mostly did.

**Comment by amymhaddad** [comment]
@paulghaddad I updated the tests in this PR. Ready for final review.

File: raindrops/test_raindrops.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+def test_number_is_0():
+    assert raindrops(0) == ""
+
+
+def test_factors_for_number_28():
+    assert raindrops(28) == "Plong"
+
+
+def test_factors_for_number_12():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: test_number_multiple_of_3

File: raindrops/test_raindrops.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    assert raindrops(12) == "Pling"
+
+
+def test_factors_for_number_10():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: test_number_multiple_of_5

File: raindrops/test_raindrops.py (hunk 8 above)
Comment (paulghaddad) [reply raindrops/test_raindrops.py]: This is not a base case, this and the next 2 are multiples of the base case. Try to communicate the meaning behind what you're testing.

File: raindrops/test_raindrops.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    assert raindrops(10) == "Plang"
+
+
+def test_number_with_no_matching_factor():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: Excellent test case and communication of meaning. 👏

File: raindrops/test_raindrops.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    assert raindrops(34) == "34"
+
+
+def test_number_with_factors_3_and_5():
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: Excellent names.

File: raindrops/test_raindrops.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
+from raindrops import raindrops
```
Comment (paulghaddad) [inline raindrops/test_raindrops.py]: Great work on the test cases; the only other valuable one is a number with factors 3, 5, and 7!

**Pull request merged**